notifier/notifier.go

```diff
@@ -350,3 +379,161 @@ func (n *Notifier) Collect(ch chan<- prometheus.Metric) {
 	ch <- n.queueLength
 	ch <- n.queueCapacity
 }
+
+// alertmanager holds all necessary information to send alerts
+// to an Alertmanager endpoint.
+type alertmanager struct {
+	plainURL string // test injection hook
+	labels   model.LabelSet
+}
+
+const pathLabel = "__alerts_path__"
+
+func (a alertmanager) url() string {
+	if a.plainURL != "" {
+		return a.plainURL
+	}
+	u := &url.URL{
+		Scheme: string(a.labels[model.SchemeLabel]),
+		Host:   string(a.labels[model.AddressLabel]),
+		Path:   string(a.labels[pathLabel]),
+	}
+	return u.String()
+}
+
+func (a alertmanager) send(ctx context.Context, c *http.Client, b []byte) error {
+	resp, err := ctxhttp.Post(ctx, c, a.url(), contentTypeJSON, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Any HTTP status 2xx is OK.
+	if resp.StatusCode/100 != 2 {
+		return fmt.Errorf("bad response status %v", resp.Status)
+	}
+	return err
+}
+
+// alertmanagerSet contains a set of Alertmanagers discovered via a group of service
+// discovery definitions that have a common configuration on how alerts should be sent.
+type alertmanagerSet struct {
+	ts     *discovery.TargetSet
+	cfg    *config.AlertmanagersConfig
+	client *http.Client
+
+	mtx sync.RWMutex
+	ams []alertmanager
+}
+
+func newAlertmanagerSet(cfg *config.AlertmanagersConfig) (*alertmanagerSet, error) {
+	client, err := retrieval.NewHTTPClient(cfg.HTTPClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	s := &alertmanagerSet{
+		client: client,
+		cfg:    cfg,
+	}
+	s.ts = discovery.NewTargetSet(s)
+
+	return s, nil
+}
+
+// Sync extracts a deduplicated set of Alertmanager endpoints from a list
+// of target groups definitions.
+func (s *alertmanagerSet) Sync(tgs []*config.TargetGroup) {
+	all := []alertmanager{}
+
+	for _, tg := range tgs {
+		ams, err := alertmanagerFromGroup(tg, s.cfg)
+		if err != nil {
+			log.With("err", err).Error("generating discovered Alertmanagers failed")
+			continue
+		}
+		all = append(all, ams...)
+	}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	// Set new Alertmanagers and deduplicate them along their unique URL.
```

Comment:
Mh, targets are purely deduped based on their label set. For Alertmanager, labels other than the magic ones are practically ignored. So the URL seemed to be the equivalent.

In general, the deduping is scoped to a single Alertmanager config/`alertmanagerSet` (similar to a scrape config). Within one there's only a single configuration of the HTTP client as we don't expose it in relabeling. So there cannot be any collisions along this dimension.